File: JZOffer_Python/rect_cover.py
# -*- coding:utf-8 -*-
'''
题目描述：
我们可以用2*1的小矩形横着或者竖着去覆盖更大的矩形。请问用n个2*1的小矩形无重叠地覆盖一个2*n的大矩形，总共有多少种方法？
'''

'''
解题思路：
https://www.nowcoder.com/questionTerminal/72a5a919508a4251859fb2cfb987a0e6
'''

class Solution:
    def rectCover(self, number):
        num = []
        num.append(0)
        num.append(1)
        num.append(2)
        for i in range(3, number + 1):
            num.append(num[i-1] + num[i-2])
        return num[number]
s = Solution()
print s.rectCover(6)

File: JZOffer_PHP/rect_cover.php
<?php
    // 运行时间：28ms 占用内存：3664k
	function rectCover($number)
	{
		if ($number <= 0){
			return 0;
		}
		if ($number == 1 || $number == 2){
			return $number;
		}
		$f1 = 1;
		$f2 = 2;
		while($number > 2){
			$temp = $f1 + $f2;
			$f1 = $f2;
			$f2 = $temp;
			$number--;
		}
		return $temp;
	}
	
	// 递归
	function rectCover($number)
	{
		if ($number <= 0){
			return 0;
		}
		if ($number == 1 || $number == 2){
			return $number;
		}
		return rectCover($number - 1) + rectCover($number - 2);
	}
	
	
?>